features/enlinkd/org.opennms.features.enlinkd.shell/src/main/java/org/opennms/features/enlinkd/shell/GenerateTopologyCommand.java

```diff
@@ -0,0 +1,155 @@
+ *
+ * OpenNMS(R) is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with OpenNMS(R).  If not, see:
+ *      http://www.gnu.org/licenses/
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.features.enlinkd.shell;
+
+import java.io.IOException;
+import java.sql.SQLException;
+
+import org.apache.karaf.shell.api.action.Action;
+import org.apache.karaf.shell.api.action.Command;
+import org.apache.karaf.shell.api.action.Option;
+import org.apache.karaf.shell.api.action.lifecycle.Reference;
+import org.apache.karaf.shell.api.action.lifecycle.Service;
+import org.opennms.enlinkd.topogen.TopologyGenerator;
+import org.opennms.enlinkd.topogen.TopologyPersister;
+import org.opennms.netmgt.dao.api.IpInterfaceDao;
+import org.opennms.netmgt.dao.api.NodeDao;
+import org.opennms.netmgt.dao.api.SnmpInterfaceDao;
+import org.opennms.netmgt.enlinkd.persistence.api.CdpElementDao;
+import org.opennms.netmgt.enlinkd.persistence.api.CdpLinkDao;
+import org.opennms.netmgt.enlinkd.persistence.api.IsIsElementDao;
+import org.opennms.netmgt.enlinkd.persistence.api.IsIsLinkDao;
+import org.opennms.netmgt.enlinkd.persistence.api.LldpElementDao;
+import org.opennms.netmgt.enlinkd.persistence.api.LldpLinkDao;
+import org.opennms.netmgt.enlinkd.persistence.api.OspfElementDao;
+import org.opennms.netmgt.enlinkd.persistence.api.OspfLinkDao;
+
+@Command(scope="enlinkd", name="generate-topology", description="Creates a linkd topology")
+@Service
+public class GenerateTopologyCommand implements Action {
+
+    @Option(name="--nodes", description="generate <N> OmnsNodes.")
+    private Integer amountNodes;
+
+    @Option(name="--elements", description="generate <N> CdpElements")
+    private Integer amountElements;
+
+    @Option(name="--links", description="generate <N> CdpLinks")
+    private Integer amountLinks;
+
+    @Option(name="--snmpinterfaces", description="generate <N> SnmpInterfaces but not more than amount nodes")
+    private Integer amountSnmpInterfaces;
+
+    @Option(name="--ipinterfaces", description="generate <N> IpInterfaces but not more than amount snmp interfaces")
+    private Integer amountIpInterfaces;
+
+    @Option(name="--topology", description="type of topology (complete | ring | random)")
+    private String topology;
+
+    @Option(name="--protocol", description="type of protocol (cdp | isis | lldp | ospf)")
+    private String protocol;
+
+    @Option(name="--delete", description = "delete existing toplogogy (all OnmsNodes, CdpElements and CdpLinks)")
```

Comment:
When we delete, we should only delete nodes and the related links that were created by this command, and not *all* of the nodes in the database.

To achieve this, we could do something like add a new category to the nodes we generate - called something like `GeneratedNode` - and then limit the deletion to nodes with this category.

Reply:
Just to clarify, if there are existing nodes and links in the database, and we run the generation command, followed by the associated delete, the database should ideally have the same state as it did *before* the generation command was ran.

Reply:
understand, I'm on it...